pkg/template/event_test.go

```diff
@@ -0,0 +1,309 @@
+			wantErr: false,
+		},
+		{
+			// name: "json subobject",
+			args:    args{event: testEventJSON, param: paramSubobjectEventPathVar},
+			want:    wantParamSubobjectEventPathVar,
+			wantErr: false,
+		},
+		{
+			// name: "error one eventpath not found",
+			args:    args{event: testEventJSON, param: paramOneBogusEventPathVar},
+			wantErr: true,
+			want:    wantParamOneBogusEventPathVar,
+		},
+		{
+			// name: "error multiple eventpath not found",
+			args:    args{event: testEventJSON, param: paramMultipleBogusEventPathVars},
+			wantErr: true,
+			want:    wantParamMultipleBogusEventPathVars,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.args.param.Value.StringVal, func(t *testing.T) {
+			got, err := applyEventToParam(tt.args.event, tt.args.param)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("applyEventToParam() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("applyEventToParam() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_ApplyEventToParams(t *testing.T) {
+	type args struct {
+		event  []byte
+		params []pipelinev1.Param
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    []pipelinev1.Param
+		wantErr bool
+	}{
+		{
+			name: "empty params",
+			args: args{
+				event:  testEventJSON,
+				params: []pipelinev1.Param{},
+			},
+			want:    []pipelinev1.Param{},
+			wantErr: false,
+		},
+		{
+			name: "one param",
+			args: args{
+				event:  testEventJSON,
+				params: []pipelinev1.Param{paramOneEventPathVar},
+			},
+			want:    []pipelinev1.Param{wantParamOneEventPathVar},
+			wantErr: false,
+		},
+		{
+			name: "multiple params",
+			args: args{
+				event: testEventJSON,
+				params: []pipelinev1.Param{
+					paramOneEventPathVar,
+					paramMultipleUniqueEventPathVars,
+					paramSubobjectEventPathVar,
+				},
+			},
+			want: []pipelinev1.Param{
+				wantParamOneEventPathVar,
+				wantParamMultipleUniqueEventPathVars,
+				wantParamSubobjectEventPathVar,
+			},
+			wantErr: false,
+		},
+		{
+			name: "error one eventpath not found",
+			args: args{
+				event: testEventJSON,
+				params: []pipelinev1.Param{
+					paramOneBogusEventPathVar,
+					paramMultipleUniqueEventPathVars,
+					paramSubobjectEventPathVar,
+				},
+			},
+			want:    nil,
+			wantErr: true,
+		},
+		{
+			name: "error multiple eventpaths not found",
+			args: args{
+				event: testEventJSON,
+				params: []pipelinev1.Param{
+					paramOneBogusEventPathVar,
+					paramMultipleBogusEventPathVars,
+				},
+			},
+			want:    nil,
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := ApplyEventToParams(tt.args.event, tt.args.params)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("ApplyEventToParams() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("ApplyEventToParams() = %v, want %v", got, tt.want)
```

Comment:
it can be really helpful to use `cmp.Diff` for this kinda assertion, so you can report exactly what was different instead of making the person running the test figure it out, e.g.:

```
if diff := cmp.Diff(got, tt.want); diff != "" {
    t.Errorf("Get PullRequest: -want +got: %s", diff) // it's helpful to include that "-want +got" bit cuz its hard to reason it out on your own
}
```

Reply:
Thanks! Good point, that'll be a lot nicer to read 👍

Reply:
Although, if I'm not mistaken, we want to do `cmp.Diff(tt.want, got)` for `-want +got`?